minitask3/akhmad/follower.py

```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/env python
+import rospy
+import cv2
+import cv_bridge
+import numpy
+
+from geometry_msgs.msg import Twist
+from sensor_msgs.msg import Image, LaserScan
+
+LOWER_GREEN = numpy.array([60, 100, 50])
+UPPER_GREEN = numpy.array([60, 255, 255])
+
+SAMPLE_WINDOW = 20  # Degrees
+
+TURN_MAX_SPEED = 0.2
+TURN_GAIN = 0.01
+
+FORWARD_SPEED = 0.3
+
+
+class Follower():
+
+    def __init__(self):
+        rospy.init_node('robot_control', anonymous=True)
+        self.vel_pub = rospy.Publisher("/cmd_vel", Twist, queue_size=1)
+        self.bridge = cv_bridge.CvBridge()
+
+        self.move_twist = Twist()
+
+        self.front = False
+        self.right = False
+        self.green_block_detected = False
+        self.centroids = []
+        self.width_middle = None
+
+    def start(self):
+        self.scan_sub = rospy.Subscriber("/scan", LaserScan, self.onscan)
+        self.camera_sub = rospy.Subscriber("/camera/rgb/image_raw",
+                                           Image, self.onimage)
+        rospy.spin()
+
+    def move(self, turnspeed, forwardspeed):
+        '''
+        Move the robot forward at forwardspeed, and turn at turnspeed.
+        '''
+        self.move_twist.linear.x = forwardspeed
+        self.move_twist.angular.z = turnspeed
+        self.vel_pub.publish(self.move_twist)
+
+    def sample_scan(self, scandata, target, window_size=SAMPLE_WINDOW):
+        '''
+        Returns the minimum value in the given window_size, centered on target
+        '''
+        assert type(window_size) == int, "window_size should be an integer"
+        offset = window_size / 2
+
+        min_val = scandata.ranges[target - offset]
+        for i in range(target - offset + 1, target + offset):
+            min_val = min(scandata.ranges[i], min_val)
+        return min_val
+
+    def move_by_image(self):
+        rospy.loginfo("Walking towards green block")
+        err_x = float('inf')
+        err_y = float('inf')
+
+        # Walk towards block that is closest to middle of image width
+        for c in self.centroids:
```

Comment:
This is a nice way of setting which block to travel to

Reply:
I just realised this may be why the robot is wobbling. If it is continuously switching between the two targets then the error will keep changing so it will bounce

Reply:
I thought about that too, but robot wobbling even when there is only one target in sight